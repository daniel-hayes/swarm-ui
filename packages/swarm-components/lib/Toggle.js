"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var React = _interopRequireWildcard(require("react"));

var _Icon = _interopRequireDefault(require("./Icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var FILLS = {
  checked: 'var(--color-viridian)',
  unchecked: 'var(--color-gray-4)'
};

var Toggle = function Toggle(props) {
  var _props$checked = props.checked,
      checked = _props$checked === void 0 ? false : _props$checked,
      _props$disabled = props.disabled,
      disabled = _props$disabled === void 0 ? false : _props$disabled,
      rest = _objectWithoutProperties(props, ["checked", "disabled"]);

  var checkedStatus = checked ? 'checked' : 'unchecked';
  var fillIcon = checked ? 'check' : 'cross';
  return React.createElement("button", _extends({
    "data-swarm-toggle": checkedStatus,
    role: "checkbox",
    type: "button",
    "aria-checked": checked,
    "aria-readonly": disabled,
    checked: checked,
    disabled: disabled
  }, rest), React.createElement("span", {
    "data-swarm-toggle-switch-disc": true
  }, React.createElement(_Icon["default"], {
    shape: fillIcon,
    size: "xs",
    color: FILLS[checkedStatus]
  })));
};

var _default = Toggle;
exports["default"] = _default;
Toggle.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "Toggle",
  "props": {
    "checked": {
      "required": true,
      "flowType": {
        "name": "boolean"
      },
      "description": "Indicated whether the toggle is selected"
    },
    "disabled": {
      "required": false,
      "flowType": {
        "name": "boolean"
      },
      "description": "Indicated whether the toggle is disabled"
    },
    "onChange": {
      "required": false,
      "flowType": {
        "name": "signature",
        "type": "function",
        "raw": "() => null",
        "signature": {
          "arguments": [],
          "return": {
            "name": "null"
          }
        }
      },
      "description": "Action to be performed when the toggle is interacted with"
    }
  }
};