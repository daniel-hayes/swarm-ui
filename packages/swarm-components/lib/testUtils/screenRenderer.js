"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _hapi = _interopRequireDefault(require("hapi"));

var _path = _interopRequireDefault(require("path"));

var _inert = _interopRequireDefault(require("inert"));

var _puppeteer = _interopRequireDefault(require("puppeteer"));

var _server = _interopRequireDefault(require("react-dom/server"));

var screenRenderer =
/*#__PURE__*/
function () {
  function screenRenderer(config) {
    (0, _classCallCheck2["default"])(this, screenRenderer);
    this.config = (0, _objectSpread2["default"])({
      viewport: {
        width: 800,
        height: 600
      },
      verbose: false,
      port: 4000,
      padding: '1em'
    }, config);
    this.routeIndex = 0;
  }

  (0, _createClass2["default"])(screenRenderer, [{
    key: "log",
    value: function log() {
      if (this.config.verbose) {
        var _console;

        (_console = console).log.apply(_console, arguments);
      }
    }
  }, {
    key: "init",
    value: function () {
      var _init = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee() {
        var _this$config, port, staticPath;

        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _this$config = this.config, port = _this$config.port, staticPath = _this$config.staticPath;
                _context.next = 3;
                return _puppeteer["default"].launch();

              case 3:
                this.browser = _context.sent;
                this.log('Browser is running');
                this.server = _hapi["default"].server({
                  port: port
                });
                _context.next = 8;
                return this.server.register(_inert["default"]);

              case 8:
                // serving static files
                this.server.route({
                  method: 'GET',
                  path: '/static/{param*}',
                  handler: {
                    directory: {
                      path: _path["default"].join(__dirname, staticPath)
                    }
                  }
                });
                _context.next = 11;
                return this.server.start();

              case 11:
                this.log('Server running on %s', this.server.info.uri);
                return _context.abrupt("return", this);

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function init() {
        return _init.apply(this, arguments);
      }

      return init;
    }()
  }, {
    key: "createRoute",
    value: function createRoute(slug, element) {
      var _this = this;

      var links = this.config.stylesheets.map(function (stylesheet) {
        return "<link\n\t\t\t\t\t\trel=\"stylesheet\"\n\t\t\t\t\t\ttype=\"text/css\"\n\t\t\t\t\t\thref=\"/static/".concat(stylesheet, "\" />");
      }).join('\n');
      return {
        method: 'GET',
        path: "/".concat(slug),
        handler: function handler() {
          return "<html>\n\t\t\t\t\t\t\t\t<head>\n\t\t\t\t\t\t\t\t\t".concat(links, "\n\t\t\t\t\t\t\t\t</head>\n\t\t\t\t\t\t\t\t<body style=\"padding: ").concat(_this.config.padding, "\">\n\t\t\t\t\t\t\t\t").concat(_server["default"].renderToStaticMarkup(element), "\n\t\t\t\t\t\t\t\t</body>\n\t\t\t\t\t\t\t</html>");
        }
      };
    }
  }, {
    key: "stop",
    value: function () {
      var _stop = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee2() {
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.browser.close();

              case 2:
                return _context2.abrupt("return", this.server.stop());

              case 3:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function stop() {
        return _stop.apply(this, arguments);
      }

      return stop;
    }()
  }, {
    key: "screenshot",
    value: function () {
      var _screenshot = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee3(element, screenshotConfig) {
        var page, slug, testUrl;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this.browser.newPage();

              case 2:
                page = _context3.sent;
                page.setViewport(screenshotConfig && screenshotConfig.viewport || this.config.viewport);
                slug = "route-".concat(this.routeIndex++);
                this.server.route(this.createRoute(slug, element));
                testUrl = "".concat(this.server.info.uri, "/").concat(slug);
                this.log("Testing: ".concat(testUrl));
                _context3.next = 10;
                return page["goto"](testUrl);

              case 10:
                return _context3.abrupt("return", page.screenshot());

              case 11:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function screenshot(_x, _x2) {
        return _screenshot.apply(this, arguments);
      }

      return screenshot;
    }()
  }]);
  return screenRenderer;
}();

var _default =
/*#__PURE__*/
function () {
  var _ref = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee4(config) {
    return _regenerator["default"].wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            return _context4.abrupt("return", new screenRenderer(config).init());

          case 1:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));

  return function (_x3) {
    return _ref.apply(this, arguments);
  };
}();

exports["default"] = _default;