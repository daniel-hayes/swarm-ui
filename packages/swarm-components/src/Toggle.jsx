import * as React from 'react';
import Icon from './Icon';

export type Props = {
  checked: boolean,
  disabled?: boolean,
  name: string,
  id: string,
  onChange?: e => null
};

const FILLS = {
  checked: 'var(--color-viridian)',
  unchecked: 'var(--color-gray-4)',
};

const Toggle = (props: Props): React.ReactElement<'span'> => {
  const { checked, disabled, id, name, ...rest } = props;
  const checkedStatus = checked ? 'checked' : 'unchecked';
  const fillIcon = checked ? 'check' : 'cross';

  return (
    <label
      data-swarm-toggle={checkedStatus}
      role='checkbox'
      aria-checked={checked}
      id={id}
      {...rest}
    >
      <input
        type='checkbox'
        checked={checked}
        disabled={disabled}
        name={name}
        id={id}
      />
      <span data-swarm-toggle-switch-disc>
        <Icon shape={fillIcon} size='xs' color={FILLS[checkedStatus]} />
      </span>
    </label>
  );
};

export default Toggle;
