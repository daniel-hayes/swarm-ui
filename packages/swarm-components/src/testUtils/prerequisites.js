import fs from 'fs';
import Path from 'path';
import { Exception } from 'handlebars';

// setup.js
module.exports = async () => {
	if (
		!fs.existsSync(
			Path.join(__dirname, '../../../swarm-styles/dist/global.css')
		)
	) {
		throw new Exception(
			"Can't find packages/swarm-styles/dist/global.css! Did you run `yarn build`?"
		);
	}

	if (
		!fs.existsSync(Path.join(__dirname, '../../../swarm-styles/dist/main.css'))
	) {
		throw new Exception(
			"Can't find packages/swarm-styles/dist/main.css! Did you run `yarn build`?"
		);
	}

	if (
		!fs.existsSync(
			Path.join(__dirname, '../../../swarm-docs/src/assets/graphik.css')
		)
	) {
		throw new Exception(
			"Can't find packages/swarm-docs/src/assets/graphik.css"
		);
	}
};
