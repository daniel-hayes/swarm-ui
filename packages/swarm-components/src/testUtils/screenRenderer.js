import Hapi from 'hapi';
import Path from 'path';
import inert from 'inert';

import puppeteer from 'puppeteer';

import ReactDOMServer from 'react-dom/server';

class screenRenderer {
	constructor(config) {
		this.config = {
			viewport: { width: 800, height: 600 },
			verbose: false,
			port: 4000,
			host: 'localhost',
			padding: '1em',
			...config,
		};

		this.routeIndex = 0;
	}

	log(...params) {
		if (this.config.verbose) {
			console.log(...params);
		}
	}

	async init() {
		const { port, host, staticPath } = this.config;

		this.browser = await puppeteer.launch();
		this.log('Browser is running');

		this.server = Hapi.server({
			port,
			host,
		});

		await this.server.register(inert);

		// serving static files
		this.server.route({
			method: 'GET',
			path: '/static/{param*}',
			handler: {
				directory: {
					path: Path.join(__dirname, staticPath),
				},
			},
		});

		await this.server.start();
		this.log('Server running on %s', this.server.info.uri);

		return this;
	}

	createRoute(slug, element, bodyStyle='') {
		const links = this.config.stylesheets
			.map(
				stylesheet =>
					`<link
						rel="stylesheet"
						type="text/css"
						href="/static/${stylesheet}" />`
			)
			.join('\n');

		return {
			method: 'GET',
			path: `/${slug}`,
			handler: () => `<html>
								<head>
									${links}
								</head>
								<body style="padding: ${this.config.padding}; ${bodyStyle}">
								${ReactDOMServer.renderToStaticMarkup(element)}
								</body>
							</html>`,
		};
	}

	async stop() {
		await this.browser.close();
		return this.server.stop();
	}

	async screenshot(element, screenshotConfig) {
		const page = await this.browser.newPage();
		page.setViewport(
			(screenshotConfig && screenshotConfig.viewport) || this.config.viewport
		);

		const slug = `route-${this.routeIndex++}`;

		this.server.route(this.createRoute(slug, element, screenshotConfig.bodyStyle));

		const testUrl = `${this.server.info.uri}/${slug}`;

		this.log(`Testing: ${testUrl}`);
		await page.goto(testUrl);

		return page.screenshot();
	}
}

export default async config => new screenRenderer(config).init();
