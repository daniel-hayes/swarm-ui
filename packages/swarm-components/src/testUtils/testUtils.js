import React from 'react';

export const getTestMarkup = (testCases) => (
    <div>
        {testCases.map((test, index) => {
            const label = test[0];
            const component = test[1];
            const wrapperStyle = test[2];

            return (
                <div key={index}>
                    <p>{label}</p>
                    <div style={wrapperStyle ? wrapperStyle : {}}>
                        {component}
                    </div>
                    <br/>
                </div>
            );
        })}
    </div>
);
