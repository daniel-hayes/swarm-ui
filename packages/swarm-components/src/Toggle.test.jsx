import React from 'react';

import { toMatchImageSnapshot } from 'jest-image-snapshot';
expect.extend({ toMatchImageSnapshot });

import getScreenRenderer from './testUtils/screenRenderer';
import Toggle from './Toggle';
import { getTestMarkup } from './testUtils/testUtils.js';

describe('Toggle', () => {
	let renderer;

	// This is ran when the suite starts up.
	beforeAll(async () => {
		renderer = await getScreenRenderer({
			viewport: { width: 200, height: 1000 },
			// verbose: true,
		});
	});

	// This is ran when the suite is done - stop the renderer.
	afterAll(() => {
		// comment next line out if you want to open it in your browser for debugging
		return renderer.stop();
	});

	const testCases = [
		['Default', <Toggle key="1" />],
		['Checked', <Toggle key="2" checked />],
		['Disabled', <Toggle key="3" disabled />],
		['Disabled Checked', <Toggle key="4" disabled checked />],
	];

	it('Should match screenshot', async () => {
		expect(
			await renderer.screenshot(getTestMarkup(testCases))
		).toMatchImageSnapshot();
	});
});
