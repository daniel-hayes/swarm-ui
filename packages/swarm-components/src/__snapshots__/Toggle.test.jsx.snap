// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Snapshot: Checked 1`] = `
<button
  aria-checked={true}
  aria-readonly={false}
  checked={true}
  data-swarm-toggle="checked"
  disabled={false}
  role="checkbox"
  type="button"
>
  <span
    data-swarm-toggle-switch-disc={true}
  >
    <Icon
      color="var(--color-viridian)"
      shape="check"
      size="xs"
    />
  </span>
</button>
`;

exports[`Toggle Snapshot: Default 1`] = `
<button
  aria-checked={false}
  aria-readonly={false}
  checked={false}
  data-swarm-toggle="unchecked"
  disabled={false}
  role="checkbox"
  type="button"
>
  <span
    data-swarm-toggle-switch-disc={true}
  >
    <Icon
      color="var(--color-gray-4)"
      shape="cross"
      size="xs"
    />
  </span>
</button>
`;

exports[`Toggle Snapshot: Disabled 1`] = `
<button
  aria-checked={false}
  aria-readonly={true}
  checked={false}
  data-swarm-toggle="unchecked"
  disabled={true}
  role="checkbox"
  type="button"
>
  <span
    data-swarm-toggle-switch-disc={true}
  >
    <Icon
      color="var(--color-gray-4)"
      shape="cross"
      size="xs"
    />
  </span>
</button>
`;

exports[`Toggle Snapshot: Disabled Checked 1`] = `
<button
  aria-checked={true}
  aria-readonly={true}
  checked={true}
  data-swarm-toggle="checked"
  disabled={true}
  role="checkbox"
  type="button"
>
  <span
    data-swarm-toggle-switch-disc={true}
  >
    <Icon
      color="var(--color-viridian)"
      shape="check"
      size="xs"
    />
  </span>
</button>
`;
